File: sdk-libs/compressed-token-sdk/src/compressible/withdraw_funding_pool.rs
use light_ctoken_types::COMPRESSED_TOKEN_PROGRAM_ID;
use solana_instruction::{AccountMeta, Instruction};
use solana_pubkey::Pubkey;

/// Creates a withdraw funding pool instruction to withdraw SOL from the pool PDA
///
/// # Arguments
/// * `pool_pda` - The pool PDA that holds the funds
/// * `pool_pda_bump` - The bump seed for the pool PDA
/// * `authority` - The authority (must be a signer and match pool PDA derivation)
/// * `destination` - The destination account to receive the withdrawn funds
/// * `amount` - The amount of lamports to withdraw
///
/// # Returns
/// The withdraw funding pool instruction
pub fn withdraw_funding_pool(
    pool_pda: Pubkey,
    pool_pda_bump: u8,
    authority: Pubkey,
    destination: Pubkey,
    amount: u64,
) -> Instruction {
    // Build instruction data: [discriminator: u8][bump: u8][amount: u64]
    let mut instruction_data = vec![105u8]; // WithdrawFundingPool instruction discriminator
    instruction_data.push(pool_pda_bump);
    instruction_data.extend_from_slice(&amount.to_le_bytes());

    let accounts = vec![
        // Pool PDA (source of funds) - must be writable
        AccountMeta::new(pool_pda, false),
        // Authority (signer) - must match pool PDA derivation
        AccountMeta::new_readonly(authority, true),
        // Destination (receives funds) - must be writable
        AccountMeta::new(destination, false),
        // System program
        AccountMeta::new_readonly(Pubkey::default(), false),
    ];

    Instruction {
        program_id: Pubkey::from(COMPRESSED_TOKEN_PROGRAM_ID),
        accounts,
        data: instruction_data,
    }
}

File: js/token-kit/src/instructions/withdraw-funding-pool.ts
/**
 * Withdraw from funding pool instruction.
 */

import type { Address } from '@solana/addresses';
import {
    AccountRole,
    type Instruction,
    type AccountMeta,
} from '@solana/instructions';

import { getU64Encoder } from '@solana/codecs';

import {
    DISCRIMINATOR,
    LIGHT_TOKEN_PROGRAM_ID,
    SYSTEM_PROGRAM_ID,
} from '../constants.js';

/**
 * Parameters for withdrawing from a funding pool.
 */
export interface WithdrawFundingPoolParams {
    /** Rent sponsor pool PDA (writable) */
    rentSponsor: Address;
    /** Compression authority (signer) */
    compressionAuthority: Address;
    /** Destination account receiving withdrawn lamports (writable) */
    destination: Address;
    /** Compressible config account (readonly) */
    compressibleConfig: Address;
    /** Amount of lamports to withdraw */
    amount: bigint;
}

/**
 * Creates a withdraw funding pool instruction (discriminator: 105).
 *
 * Withdraws lamports from the rent sponsor funding pool.
 *
 * Account layout:
 * 0: rent_sponsor (writable) - Pool PDA
 * 1: compression_authority (signer)
 * 2: destination (writable) - Receives withdrawn lamports
 * 3: system_program (readonly)
 * 4: compressible_config (readonly)
 *
 * @param params - Withdraw funding pool parameters
 * @returns The withdraw funding pool instruction
 */
export function createWithdrawFundingPoolInstruction(
    params: WithdrawFundingPoolParams,
): Instruction {
    const {
        rentSponsor,
        compressionAuthority,
        destination,
        compressibleConfig,
        amount,
    } = params;

    // Build accounts
    const accounts: AccountMeta[] = [
        { address: rentSponsor, role: AccountRole.WRITABLE },
        { address: compressionAuthority, role: AccountRole.READONLY_SIGNER },
        { address: destination, role: AccountRole.WRITABLE },
        { address: SYSTEM_PROGRAM_ID, role: AccountRole.READONLY },
        { address: compressibleConfig, role: AccountRole.READONLY },
    ];

    // Build instruction data: discriminator (u8) + amount (u64)
    const amountBytes = getU64Encoder().encode(amount);
    const data = new Uint8Array(1 + amountBytes.length);
    data[0] = DISCRIMINATOR.WITHDRAW_FUNDING_POOL;
    data.set(new Uint8Array(amountBytes), 1);

    return {
        programAddress: LIGHT_TOKEN_PROGRAM_ID,
        accounts,
        data,
    };
}
